### PR TITLE
Removing `m_hostAssemblyMap` and ceremony around it

### PR DESCRIPTION
**Pull request by VSadov**
Since we have only one AppDomain, there is no point in having a per-AppDomain {`HostAssembly`-->`DomainAssembly`} map.

Removing this map allows to remove couple locks, including the one taken during GC stack walks, and coordination with EE suspension.

Also removes `SHash` support for callouts to host in hosted scenario (comments say "hosted", I assume they mean "SQL-hosted"). 
I do not believe such scenarios exist anymore.

### CODE REVIEW

**Comment by VSadov** [comment]
Fairly simple change - mostly removing code.

**Review by jkotas — APPROVED** [review]
Thanks!

File: src/coreclr/vm/domainfile.h [hunk 1]
```diff
@@ -587,6 +587,8 @@ class DomainAssembly : public DomainFile
     void Allocate();
     void DeliverSyncEvents();
     void DeliverAsyncEvents();
+    void RegisterWithHostAssembly();
+    void UnRegisterFromHostAssembly();
```
Comment (elinor-fung) [inline src/coreclr/vm/domainfile.h:591]: `UnRegister` -> `Unregister`

File: src/coreclr/inc/shash.h [hunk 2]
```diff
@@ -278,7 +273,7 @@ class EMPTY_BASES_DECL SHash : public TRAITS
 
     // Utility function to allocate new table (does not copy the values into it yet). Returns the size of new table in
     // *pcNewTableSize (finds next prime).
-    // Phase 1 of code:Reallocate - it is split to support code:AddPhases.
+    // Phase 1 of code:Reallocate
     element_t * AllocateNewTable(count_t requestedSize, count_t * pcNewTableSize);
```
[the remark this answers is not in the document]
Comment (VSadov) [reply src/coreclr/inc/shash.h:277]: They are not small/trivial methods and have different responsibilities, so I figured it is ok to leave them as separate.

[the remark this answers is not in the document]
Comment (VSadov) [reply src/coreclr/inc/shash.h:277]: There are also other callers - like `AllocateNewTable` is also called by `Grow_OnlyAllocateNewTable`.

[the remark this answers is not in the document]
Comment (VSadov) [reply src/coreclr/inc/shash.h:277]: Others have multiple callers too.

**Comment by VSadov** [comment]
Thanks!!

**Review by elinor-fung — APPROVED** [review]

File: src/coreclr/vm/domainfile.h (hunk 1 above)
Comment (VSadov) [reply src/coreclr/vm/domainfile.h:591]: Thanks! I will do the rename in some other change.

**Pull request merged**